File: react-native/react-native-totvs-clockin-vision/ios/Core/RecognitionModel.swift
//
//  DetectionModel.swift
//  TOTVSClockInVision
//
//  Created by Jansel Rodriguez on 7/30/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation

/**
* Model as representative of model classes that are capable of detection
* tasks over an input value.
*/
public protocol RecognitionModel : Model  {
    associatedtype Input
    associatedtype Output
    
    /**
     * Perform detection over the input producing a list of detected entities of
     * type [Output]. Optionally also perform detection on the provided input
     *
     * This method might throw [IllegalStateException] if called before being the model
     * trained. This is implementation detail
     *
     * @discussion requires to run on a working thread.
     */
    func recognize(
        input: Input,
        includeDetection: Bool,
        onRecognized: @escaping (ModelOutput<Output>) -> Void
    ) throws
}

File: react-native/react-native-totvs-clockin-vision/android/src/main/java/com/totvs/clockin/vision/core/RecognitionModel.kt
package com.totvs.clockin.vision.core

import androidx.annotation.WorkerThread

/**
 * Model as representative of model classes that are capable of recognition
 * tasks over an input value.
 */
interface RecognitionModel<in Input, out Output> : Model {
    /**
     * Perform recognition over the input producing a list of recognized entities of
     * type [Output]. Optionally also perform detection on the provided input
     *
     * This method might throw [IllegalStateException] if called before being the model
     * trained. This is implementation detail
     */
    @WorkerThread
    @Throws(IllegalStateException::class)
    fun recognize(
        input: Input,
        includeDetection: Boolean = true,
        onRecognized: (ModelOutput<Output>) -> Unit
    )
}